lang/lvl_is_lex_preserve/lvl_is_lex.c

```diff
@@ -0,0 +1,316 @@
+        return 1+lvl_is_lex_dist_to_vowel(val_cdr(rest));
+}
+
+static const char * const lvl_is_lex_onset_trigrams[] = {
+    // TODO adapt to is
+    "str", "spy", "spr", "spl", "sky", "skw", "skr", "skl", NULL
+};
+static const char * const lvl_is_lex_onset_bigrams[] = {
+    // TODO adapt to is
+    "sv", "sj", "sr",
+    "pv", "pj", "pr", 
+    "tv", "tj", "tr",
+    "kv", "kj", "kr",
+    "fr",
+
+
+    "zw", "zl",
+    "vy", "vr", "vl",
+    "thw", "thr",
+    "ty", "tw",
+    "tr", /* "ts", */
+    "shw", "shr", "shn", "shm", "shl",
+    "sw", "sv", "st", "sr", "sp", "sn", "sm", "sl", "sk", "sf",
+    "py", "pw", "pr", "pl",
+    "ny",
+    "my", "mr",
+    "ly",
+    "ky", "kw", "kr", "kl",
+    "hhy", "hhw", "hhr", "hhl",
+    "gy", "gw", "gr", "gl", 
+    "fy", "fr", "fl", 
+    "dy", "dw", "dr",
+    "by", "bw", "br", "bl",
+    NULL
+};
+
+static int lvl_is_lex_onset_bigram(const cst_val *rest)
+{
+    // TODO adapt to is
+    char x[10];
+    int i;
+
+    cst_sprintf(x,"%s%s",val_string(val_car(rest)),
+           val_string(val_car(val_cdr(rest))));
+    for (i=0; lvl_is_lex_onset_bigrams[i]; i++)
+        if (cst_streq(x,lvl_is_lex_onset_bigrams[i]))
+            return TRUE;
+    return FALSE;
+}
+
+static int lvl_is_lex_onset_trigram(const cst_val *rest)
+{
+    // TODO adapt to is
+    char x[15];
+    int i;
+
+    cst_sprintf(x,"%s%s%s",val_string(val_car(rest)),
+           val_string(val_car(val_cdr(rest))),
+           val_string(val_car(val_cdr(val_cdr(rest)))));
+    for (i=0; lvl_is_lex_onset_trigrams[i]; i++)
+        if (cst_streq(x,lvl_is_lex_onset_trigrams[i]))
+            return TRUE;
+    return FALSE;
+}
+
+int lvl_is_syl_boundary_mo(const cst_item *i,const cst_val *rest)
+{
+    // TODO adapt to is
+    /* syl boundary maximal onset */
+    int d2v;
+
+    if (rest == NULL)
+	return TRUE;
+    else if (lvl_is_is_silence(val_string(val_car(rest))))
+	return TRUE;
+    else if (!lvl_is_has_vowel_in_list(rest)) 
+        /* no more vowels so rest *all* coda */
+	return FALSE;
+    else if (!lvl_is_has_vowel_in_syl(i))  /* need a vowel */
+        /* no vowel yet in syl so keep copying */
+	return FALSE;
+    else if (lvl_is_is_vowel(val_string(val_car(rest))))
+        /* next is a vowel, syl has vowel, so this is a break */
+	return TRUE;
+    else if (cst_streq("ng",val_string(val_car(rest))))
+        /* next is "ng" which can't start a word internal syl */
+	return FALSE;
+    else 
+    {
+        /* want to know if from rest to the next vowel is a valid onset */
+        d2v = lvl_is_lex_dist_to_vowel(rest);
+        if (d2v < 2)
+            return TRUE;
+        else if (d2v > 3)
+            return FALSE;
+        else if (d2v == 2) 
+            return lvl_is_lex_onset_bigram(rest);
+        else /* if (d2v == 3) */
+            return lvl_is_lex_onset_trigram(rest);
+        return TRUE;
+    }
+
+}
+
+cst_utterance *lvl_is_lex_postlex(cst_utterance *u)
+{
+    return u;
+}
+
+cst_lexicon lvl_is_lex;
```

Comment:
I understand that `lvl_is_lex` is initialized by `lvl_is_lex_init()`, but is it referred to by its name outside of this module ? If not, you can make it static.

Reply:
I can't speak to exactly how this is referred to in other places, but here are the main references to it I could find with a quick search in the voice's flite subdirectory:
In `flite_lang_list.c`
```
#include "flite.h"

void lvl_is_lang_init(cst_voice *v);
cst_lexicon *lvl_is_lex_init(void);


void flite_set_lang_list(void)
{
   flite_add_lang("lvl_is_lang",lvl_is_lang_init,lvl_is_lex_init);
}
```
in `lvl_is_f.c`:
```
    cst_lexicon *lex;
...
    /* Lexicon */
    lex = lvl_is_lex_init();
    flite_feat_set(v->features,"lexicon",lexicon_val(lex));
    flite_feat_set(v->features,"postlex_func",uttfunc_val(lex->postlex));
```
I'll happily make it static if you think that is an improvement.

Reply:
yeah you can make it static, it's only accessed by `lvl_is_lex_init()`

Reply:
Just to clarify, are we talking about moving the declaration inside the function or simply adding the static keyword to the declaration?

Reply:
Just add `static` in front of it. This makes it module-local, i.e. it can never be access from outside this file. Otherwise it's a global variable. In C++, one would encapsulate in an anonymous namespace to achieve the same effect, like that:

``c++
namespace {
    cst_lexicon *lex;
}
```

But Flite is totally based on C.